fix(renderer): extruded polygon in Feature2Mesh

zarov opened the pull request:
Simple polygons in Feature2Mesh were not extruded properly. This has
been resolved, and in the same time the two very similar part of the
code related to this have been merged.

(moved out of #756)

zarov committed:
fix(vector): extruded polygon in Feature2Mesh

Simple polygons in Feature2Mesh were not extruded properly. This has
been resolved, and in the same time the two very similar part of the
code related to this have been merged.

A test verifying the correct extrusion result has been added.

peppsac commented:
> Simple polygons in Feature2Mesh were not extruded properly

Could you add a unit test that would fail before this PR and gets fixed with this PR?

Maybe something that would:
- parse a simple geojson with a single polygon (a square)
- convert it to Mesh without extrusion and verify the number of vertices (should be 4)
- convert it to Mesh with extrusion and verify the number of vertices (should be 8)

What do you think?

zarov commented:
Yep, I'll do it !

zarov commented:
@peppsac is the test I added ok for you ?

peppsac commented on test/feature2mesh_unit_test.js:
```diff
@@ -55,4 +55,19 @@ describe('Feature2Mesh', function () {
+                extrude: 2,
+            })(feature[0]);
+
+            const indexExtrude = new Uint16Array([5, 4, 7, 7, 6, 5, 0, 4, 1, 1,
```
You're hardcoding in the test the ordering of the vertices. IMHO this isn't a good idea, because the test might fail if this order change - even if the output is correct.

peppsac commented on test/feature2mesh_unit_test.js:
```diff
@@ -55,4 +55,19 @@ describe('Feature2Mesh', function () {
+
+            assert.equal(noExtrude.geometry.attributes.position.count, 4);
+            assert.equal(extrude.geometry.attributes.position.count, 8);
+            assert.equal(extrude.geometry.index.array.every((e, i) => indexExtrude[i] === e), true);
```
You can verify the number of triangles instead.

It's less likely to change: it should be 2 triangles per face of the quad/cube. So 2 triangles for the `noExtrude` case and 6*2 triangles for the `extrude`  one.

zarov replied:
This makes more sense than verifying vertices, indeed.

peppsac reviewed (commented):
One small comment + I'd change the scope use in the commit to `vector`.

zarov commented:
So I updated the test, and it happened that the extrusion was wrong on paper (but good on visuals ?), so I added a fix to it, but I'm really lost here as we already have a good visual on example like `wfs`. Let me know if I missed something here.

peppsac commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -252,12 +257,16 @@ function geometryToExtrudedPolygon(geometry, properties, options, multiGeomAttri
 
         // triangulate the top face
-        const start = vertices.length + polygonGeometry.contour.offset + offset;
+        let start = vertices.length + polygonGeometry.contour.offset + offset;
```
Any idea why `end` doesn't depend on ` polygonGeometry.contour.offset`? (or why end isn't `start + count`)?

zarov replied:
I concede that I didn't think here, but it could make more sense (and more understandable).

peppsac commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -115,6 +115,11 @@ function addExtrudedPolygonSideFaces(indices, length, ring, isClockWise) {
         }
     }
+
+    // last indexes are too high - wrap them inside boundaries
```
This is ugly :)

Maybe use the modulo in the previous loop, for all `i + ...` cases.
Something like: `mod = ring.offset + ring.count + length;`
Then:
```js
...
indices.push((i + length) % mod);
indices.push(i);
indices.push((i + length + 1) % mod);
...
```

zarov replied:
Yes, but if we do the modulo operation each time on a big ring, we compute it for nothing. But I guess we won't have complex shapes, so I'm okay with changing it.

peppsac commented:
Did you test your change on a complex polygon? For instance the one from issue #688?

zarov commented:
I can't make the linked polygon work as a geometry simply, so I stopped for now on this. However, tests with not so complex polygon but still complex, in the WFS example, seems to give the same result as what is currently on the master. I'm working on the loading of unique geojson file as a geometry in #766.

zarov commented:
I succeeded to display the multipolygon, unfortunately, as you can see below, this PR is breaking things, so I'll be working a bit more on this.

![image](https://user-images.githubusercontent.com/741255/41278026-22189522-6e29-11e8-8160-6ace401dddf4.png)

peppsac commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -247,6 +248,46 @@ function geometryToExtrudedPolygon(geometry, properties, options, multiGeomAttri
+        const triangles = Earcut(geom.attributes.position.array.slice(start * 3,
+            start == end ? undefined : end * 3), holesOffsets, 3);
+
+        if (start == end) {
```
Why would this happen?

zarov replied:
In single polygon case, the count is 0, which is the same as `start == end`. Could be better to change it to `count == 0`, my bad.

peppsac replied:
I'd prefer `count` to have a meaningful value in all case instead :)

So the calling site should be replaced by `addPolygon(geometry, isClockWise, 0, geometry.contour.count);`

zarov committed:
fix multipolygon

peppsac commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -96,24 +96,27 @@ function addExtrudedPolygonSideFaces(indices, length, ring, isClockWise) {
     for (let i = ring.offset; i < ring.offset + ring.count; ++i) {
+        const i1 = Math.max((i + 1) % mod, ring.offset);
+        const i2 = Math.max((i + length) % mod, ring.offset);
+        const i3 = Math.max((i + length + 1) % mod, ring.offset);
```
Suggestion: `i` and `i1` could be `floor1`/`floor2` and `i2/i3` could be `ceil1` `ceil2` (or `roof1`..) ?

zarov replied:
Thanks ! Even maybe `bottom1/2` and `top1/2` as floor/ceil/roof is vocabulary for buildings.

zarov commented:
Closing this PR as everything should be fixed in #756
